server/src/main/java/org/elasticsearch/indices/ShardLimitService.java

```diff
@@ -0,0 +1,135 @@
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.ValidationException;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.Index;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.elasticsearch.cluster.metadata.IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING;
+
+public class ShardLimitService {
+    public static final Setting<Integer> SETTING_CLUSTER_MAX_SHARDS_PER_NODE =
+        Setting.intSetting("cluster.max_shards_per_node", 1000, 1, Setting.Property.Dynamic, Setting.Property.NodeScope);
+    protected final AtomicInteger shardLimitPerNode = new AtomicInteger();
+
+    public ShardLimitService(final Settings settings, ClusterService clusterService) {
+        this.shardLimitPerNode.set(SETTING_CLUSTER_MAX_SHARDS_PER_NODE.get(settings));
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(SETTING_CLUSTER_MAX_SHARDS_PER_NODE, this::setShardLimitPerNode);
+    }
+
+    // Protected for testing purposes
+    protected void setShardLimitPerNode(int newValue) {
+        this.shardLimitPerNode.set(newValue);
+    }
+
+    public int getShardLimitPerNode() {
+        return shardLimitPerNode.get();
+    }
+
+    /**
+     * Checks whether an index can be created without going over the cluster shard limit.
+     *
+     * @param settings       the settings of the index to be created
+     * @param state          the current cluster state
+     * @throws ValidationException if creating this index would put the cluster over the cluster shard limit
+     */
+    public void validateShardLimit(final Settings settings, final ClusterState state) {
+        final int numberOfShards = INDEX_NUMBER_OF_SHARDS_SETTING.get(settings);
+        final int numberOfReplicas = IndexMetadata.INDEX_NUMBER_OF_REPLICAS_SETTING.get(settings);
+        final int shardsToCreate = numberOfShards * (1 + numberOfReplicas);
+
+        final Optional<String> shardLimit = checkShardLimit(shardsToCreate, state);
+        if (shardLimit.isPresent()) {
+            final ValidationException e = new ValidationException();
+            e.addValidationError(shardLimit.get());
+            throw e;
+        }
+    }
+
+    /**
+     * Validates whether a list of indices can be opened without going over the cluster shard limit.  Only counts indices which are
+     * currently closed and will be opened, ignores indices which are already open.
+     *
+     * @param currentState The current cluster state.
+     * @param indices The indices which are to be opened.
+     * @throws ValidationException If this operation would take the cluster over the limit and enforcement is enabled.
+     */
+    public void validateShardLimit(ClusterState currentState, Index[] indices) {
```

Comment:
Super minor, but I was confused until I read the docs, maybe rename the parameter "indicesToBeOpened"?

Reply:
👍 This was moved from another class which had clearer context for this, I'll update the name.